core/consensus/babe/src/lib.rs

```diff
@@ -639,6 +645,34 @@ impl<B: Block, C> Verifier<B> for BabeVerifier<C> where
 					auxiliary: Vec::new(),
 					fork_choice: ForkChoiceStrategy::LongestChain,
 				};
+				const INVERSE_NANO: u32 = 1000_000_000;
+				let mut timestamps = self.timestamps.lock();
+				// Remainder of this block is a critical section.
+				let num_timestamps = timestamps.1.len();
+				if num_timestamps as u64 >= self.config.0.median_required_blocks {
+					let mut new_list: Vec<_> = timestamps.1.iter().map(|&(t, sl)| {
+							let offset: u128 = u128::from(self.config.get())
+								.checked_mul(1_000_000u128) // self.config.get() returns *milliseconds*
+								.and_then(|x| x.checked_mul(u128::from(slot_num).saturating_sub(u128::from(sl))))
+								.expect("we cannot have timespans long enough for this to overflow; qed");
+							let nanos = (offset % u128::from(INVERSE_NANO)) as u32;
+							let secs = (offset / u128::from(INVERSE_NANO)) as u64;
+							t + Duration::new(secs, nanos)
```

Comment:
hrm, why not just use [Duration::from_millis](https://doc.rust-lang.org/std/time/struct.Duration.html#method.from_millis) and skip converting it to `nanos`?